Change Dynamo's custom ops warning message to be less spammy

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128456

This is a short-term fix (for 2.4). In the longer term we should
fix https://github.com/pytorch/pytorch/issues/128430

The problem is that warnings.warn that are inside Dynamo print
all the time. Python warnings are supposed to print once, unless their
cache is reset: Dynamo ends up resetting that cache everytime it runs.

As a workaround we provide our own warn_once cache that is keyed on the
warning msg. I am not worried about this increasing memory usage because
that's effectively what python's warnings.warn cache does.

Test Plan:
- fix tests.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

zou3519 committed:
Change Dynamo's custom ops warning message to be less spammy

This is a short-term fix (for 2.4). In the longer term we should
fix https://github.com/pytorch/pytorch/issues/128430

The problem is that warnings.warn that are inside Dynamo print
all the time. Python warnings are supposed to print once, unless their
cache is reset: Dynamo ends up resetting that cache everytime it runs.

As a workaround we provide our own warn_once cache that is keyed on the
warning msg. I am not worried about this increasing memory usage because
that's effectively what python's warnings.warn cache does.

Test Plan:
- fix tests.

[ghstack-poisoned]

zou3519 committed:
Update on "Change Dynamo's custom ops warning message to be less spammy"


This is a short-term fix (for 2.4). In the longer term we should
fix https://github.com/pytorch/pytorch/issues/128430

The problem is that warnings.warn that are inside Dynamo print
all the time. Python warnings are supposed to print once, unless their
cache is reset: Dynamo ends up resetting that cache everytime it runs.

As a workaround we provide our own warn_once cache that is keyed on the
warning msg. I am not worried about this increasing memory usage because
that's effectively what python's warnings.warn cache does.

Test Plan:
- fix tests.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

williamwen42 reviewed (commented):
There's duplicate functionality that we can merge: https://github.com/pytorch/pytorch/blame/5d8c7f39d46699d8f8e92512309ea3499a29c08a/torch/_dynamo/utils.py#L371

zou3519 commented:
> https://github.com/pytorch/pytorch/blame/5d8c7f39d46699d8f8e92512309ea3499a29c08a/torch/_dynamo/utils.py#L371

@williamwen42 as far as I can tell this is different -- it looks like it duplicate graph breaks for torch._dynamo.explain: https://www.internalfb.com/code/fbsource/[e28a637f7eae38b04cb0eb7ee5d6bf1ea62cd4e7]/fbcode/caffe2/torch/_dynamo/symbolic_convert.py?lines=524-529

XuehaiPan commented on torch/_dynamo/utils.py:
```diff
@@ -2748,3 +2749,18 @@ def __init__(self, s):
         return self.s
+
+
+warn_once_cache: Set[str] = set({})
```
```suggestion
warn_once_cache: Set[str] = set()
```

zou3519 committed:
Update on "Change Dynamo's custom ops warning message to be less spammy"


This is a short-term fix (for 2.4). In the longer term we should
fix https://github.com/pytorch/pytorch/issues/128430

The problem is that warnings.warn that are inside Dynamo print
all the time. Python warnings are supposed to print once, unless their
cache is reset: Dynamo ends up resetting that cache everytime it runs.

As a workaround we provide our own warn_once cache that is keyed on the
warning msg. I am not worried about this increasing memory usage because
that's effectively what python's warnings.warn cache does.

Test Plan:
- fix tests.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

williamwen42 commented:
@zou3519 hmm I still think `DuplicateWarningChecker` and `warn_once` are trying to accomplish fairly similar things - not a huge deal if we don't merge functionality though.

zou3519 commented:
@pytorchbot merge -f "failures look unrelated"
